File: app/src/api/snapshots.ts
// Workspace checkpoint API.
//
// Per-turn shadow-git snapshots of the user's workspace, stored under
// `~/.yiyi/checkpoints/<workspace_hash>/git/` so the user's real `.git`
// is never touched. Agent edits can be rolled back without disturbing
// the user's own version control.

import { invoke } from "@tauri-apps/api/core";

export interface CheckpointInfo {
  session_id: string;
  turn_index: number;
  phase: "pre" | "post" | string;
  commit: string;
  parent_commit: string | null;
  created_at_ms: number;
  files_changed: number;
  insertions: number;
  deletions: number;
  changed_files: string[];
}

export interface CheckpointRestoreReport {
  restored_files: string[];
  removed_files: string[];
  /** Auto-stash commit oid if hand-edits were captured before restore. */
  stash_commit: string | null;
}

export interface FileDiff {
  path: string;
  status: "added" | "modified" | "deleted" | "renamed" | "copied";
  additions: number;
  deletions: number;
  patch: string;
  truncated: boolean;
}

export async function listCheckpoints(sessionId: string): Promise<CheckpointInfo[]> {
  return invoke<CheckpointInfo[]>("list_checkpoints", { sessionId });
}

export async function previewCheckpoint(
  sessionId: string,
  turnIndex: number,
  phase: "pre" | "post",
): Promise<FileDiff[]> {
  return invoke<FileDiff[]>("preview_checkpoint", { sessionId, turnIndex, phase });
}

export async function restoreCheckpoint(
  sessionId: string,
  turnIndex: number,
  phase: "pre" | "post",
  paths?: string[],
): Promise<CheckpointRestoreReport> {
  return invoke<CheckpointRestoreReport>("restore_checkpoint", {
    sessionId,
    turnIndex,
    phase,
    paths: paths ?? null,
  });
}

export async function discardCheckpointBranch(
  sessionId: string,
  keepThroughTurn: number,
): Promise<number> {
  return invoke<number>("discard_checkpoint_branch", {
    sessionId,
    keepThroughTurn,
  });
}

File: app/src-tauri/src/commands/snapshots.rs
//! Tauri commands for workspace checkpoints (shadow-git via libgit2).

use tauri::State;

use crate::engine::checkpoint::{self, CheckpointInfo, FileDiff, Phase};
use crate::state::AppState;

#[tauri::command]
pub async fn list_checkpoints(
    state: State<'_, AppState>,
    session_id: String,
) -> Result<Vec<CheckpointInfo>, String> {
    let workspace = state.user_workspace();
    Ok(checkpoint::list_snapshots(&session_id, &workspace))
}

#[tauri::command]
pub async fn preview_checkpoint(
    state: State<'_, AppState>,
    session_id: String,
    turn_index: u32,
    phase: Phase,
) -> Result<Vec<FileDiff>, String> {
    let workspace = state.user_workspace();
    checkpoint::preview_diff(&session_id, turn_index, phase, &workspace).await
}

#[tauri::command]
pub async fn restore_checkpoint(
    state: State<'_, AppState>,
    session_id: String,
    turn_index: u32,
    phase: Phase,
    paths: Option<Vec<String>>,
) -> Result<checkpoint::RestoreReport, String> {
    let workspace = state.user_workspace();
    let path_bufs = paths.map(|v| v.into_iter().map(std::path::PathBuf::from).collect());
    checkpoint::restore(&session_id, turn_index, phase, &workspace, path_bufs).await
}

#[tauri::command]
pub async fn discard_checkpoint_branch(
    state: State<'_, AppState>,
    session_id: String,
    keep_through_turn: u32,
) -> Result<u32, String> {
    let workspace = state.user_workspace();
    checkpoint::discard_branch_after(&session_id, keep_through_turn, &workspace).await
}
